bokeh/server/session.py

```diff
@@ -179,26 +180,6 @@ def _document_changed(self, event):
         may_suppress = self._current_patch is not None and \
                        self._current_patch.should_suppress_on_change(event)
 
-        if isinstance(event, SessionCallbackAdded):
-            if isinstance(event.callback, PeriodicCallback):
-                self._add_periodic_callback(event.callback)
-            elif isinstance(event.callback, TimeoutCallback):
-                self._add_timeout_callback(event.callback)
-            else:
-                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
-
-            return
-
-        elif isinstance(event, SessionCallbackRemoved):
-            if isinstance(event.callback, PeriodicCallback):
-                self._remove_periodic_callback(event.callback)
-            elif isinstance(event.callback, TimeoutCallback):
-                self._remove_timeout_callback(event.callback)
-            else:
-                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
-
-            return
-
         if self._pending_writes is None:
             raise RuntimeError("_pending_writes should be non-None when we have a document lock, and we should have the lock when the document changes")
 
```

Comment:
you don't need an empty one (since dispatch checks hasattr).

If it were a private API I agree don't do unneeded things. If it's public then it should be "complete" I think. The rule could be one callback matching each class in the event inheritance hierarchy, with the callback named after the class, for example.